### PR TITLE
Add API for mods to interact with other mods' preloads

### PR DESCRIPTION
**Pull request by flibber-hk**
TL;DR: This is the cleanest way to fix a bug in Custom Knight.

The rationale for this PR is as follows. Custom Knight has a feature where objects not associated with the knight can get skinned - for example, a skin might want to replace all copies of a grub with a custom sprite.

For base game items, the way this works is that the skin author declares that hash `568482B2D16BF50AE18B394EC15B290048B75DCD` is associated with the grub skin, and then CustomKnight looks up in a precomputed lookup [here](https://github.com/PrashantMohta/HollowKnight.CustomKnight/blob/aee70e3174b66036981adc2096240ae70667ddfc/CustomKnight/Res/hashcache.json#L1035) which scenes/objects need to be skinned.

Obviously this approach doesn't work if another mod has preloaded and instantiated grubs, because they won't be found in any of the listed scenes/paths.

The best approach we could come up with was for Custom Knight to add a component to preloaded objects to indicate where they originally came from, so for example if a mod preloaded a grub [like this](https://github.com/homothetyhk/HollowKnight.ItemChanger/blob/master/ItemChanger/Internal/Preloaders/GrubPreloader.cs#L7) then the data could be added as a component on the preloaded object to indicate the source of the preload (or similar). This method is added as an alternative to CK having to IL-Hook internal MAPI methods in the mod constructor.

---

Note - the new API is added as a virtual method rather than an event because the event would have to be subscribed before mod.Initialize, which is not ideal.

### CODE REVIEW

**Comment by PrashantMohta** [comment]
Idm this not being implemented in the API. It's not in the scope of CK to skin other mods' content. It's the other mods responsibility to skin its own content.

Scope of CK is to skin vanilla content, anything else is not in the scope of CK, CK is willing to provided APIs for mods if they want their objects to be skinned in the same way CK skins vanilla content. If other mods refuse to implement it, it's not CKs responsibility.

**Comment by BadMagic100** [comment]
> A feature request that is hell bent on making it CKs responsibility to skin another mods' content.
Is pretty unnecessarily aggressive and not really based in fact - the folks who believe this is a CK issue (myself includes) are not asserting that CK should be skinning all other mods' content by default, we are asserting that CK should be skinning all vanilla game sprites, without respect to what/where the GameObject is or who created it.

**Comment by PrashantMohta** [comment]
This whole thing has been unnecessarily agressive IMO 🙂 at this point, I'm just returning the favor and defending myself.
